### ISSUE TITLE
Using BodyMatchers with sub-arrays results an isEmpty being generated

### ISSUE THREAD

**Issue by mystical44** [spring-cloud/spring-cloud-contract#797]
Bug Report:

This may be related to #391...

Issue:  If you have a sub-array with multiple elements and you want to use BodyMatchers, the code tries to remove any duplicate matchers from the response body first and replace it with the bodyMatcher check into the output test file.  

When you add the final element body matcher in the contract, the code then tries to remove the final response body check, leaving it empty essentially. A subsequent process that examines the remaining response body checks then sees this array as empty and thinks that it needs to add an isEmpty check as a result.

Ultimately, my goal is to be able to validate each occurrence of sub-array elements in the JSON file.  Perhaps there is a better way?

Example:

Contract:
```
Contract.make {
    name("get_new_toy_specs")
    request {
        description("""
            Given: A new toy request is submitted
            When: I receive the response
            Then: I would receive the toy specs
        """)
        method 'GET'
        urlPath ('/toys') {
                queryParameters {
                        parameter 'uuid': 'd4d724c4-e36e-4fd2-9baa-af7f5df17399'
                    }
        }
    }
    response {
      status 200
      body([
            toyUuid: "d4d724c4-e36e-4fd2-9baa-af7f5df17399",
            toyDescription: [
                name: "Super Whiz Bang Toy",
                stockNum: 1234,
                manufacturer: "Toy Comp",
            ],
            toyDetails: [
                [ 
                    inventory: 42,
                    description: "Toy of the year!!",
                    dimensions: [  
                        height: 45.8,
                        weight: 12.3,
                        width: 8.6,
                        length: 9.3
                    ]
                ]
            ]
      ])
      bodyMatchers {  
           //toyDescription checks
           jsonPath('$.toyDetails[*].dimensions.weight', byRegex(nonBlank()))
           jsonPath('$.toyDetails[*].dimensions.width', byRegex(nonBlank()))
           jsonPath('$.toyDetails[*].dimensions.length', byRegex(nonBlank()))
       }
      headers {
            contentType(applicationJson())
      }
    }
}
```

Results in:

```
@Test
	public void validate_get_new_toy_specs() throws Exception {
		// given:
			MockMvcRequestSpecification request = given();
		// when:
			ResponseOptions response = given().spec(request)
					.queryParam("uuid","d4d724c4-e36e-4fd2-9baa-af7f5df17399")
					.get("/toys");
		// then:
			assertThat(response.statusCode()).isEqualTo(200);
			assertThat(response.header("Content-Type")).matches("application/json.*");
		// and:
			DocumentContext parsedJson = JsonPath.parse(response.getBody().asString());
			assertThatJson(parsedJson).field("['toyUuid']").isEqualTo("d4d724c4-e36e-4fd2-9baa-af7f5df17399");
			assertThatJson(parsedJson).array("['toyDetails']").contains("['inventory']").isEqualTo(42);
			assertThatJson(parsedJson).field("['toyDescription']").field("['stockNum']").isEqualTo(1234);
			assertThatJson(parsedJson).array("['toyDetails']").contains("['description']").isEqualTo("Toy of the year!!");
			assertThatJson(parsedJson).array("['toyDetails']").field("['dimensions']").field("['height']").isEqualTo(45.8);
			assertThatJson(parsedJson).field("['toyDescription']").field("['name']").isEqualTo("Super Whiz Bang Toy");
			assertThatJson(parsedJson).field("['toyDescription']").field("['manufacturer']").isEqualTo("Toy Comp");
		// and:
			assertThat((java.lang.Iterable) parsedJson.read("$.toyDetails[*].dimensions.weight", java.util.Collection.class)).as("$.toyDetails[*].dimensions.weight").allElementsMatch("^\\s*\\S[\\S\\s]*");
			assertThat((java.lang.Iterable) parsedJson.read("$.toyDetails[*].dimensions.width", java.util.Collection.class)).as("$.toyDetails[*].dimensions.width").allElementsMatch("^\\s*\\S[\\S\\s]*");
			assertThat((java.lang.Iterable) parsedJson.read("$.toyDetails[*].dimensions.length", java.util.Collection.class)).as("$.toyDetails[*].dimensions.length").allElementsMatch("^\\s*\\S[\\S\\s]*");
	}
```

If you add the last "jsonPath('$.toyDetails[*].dimensions.height', byRegex(nonBlank()))" check to the contract:

```
      bodyMatchers {  
           //toyDescription checks
           jsonPath('$.toyDetails[*].dimensions.height', byRegex(nonBlank()))
           jsonPath('$.toyDetails[*].dimensions.weight', byRegex(nonBlank()))
           jsonPath('$.toyDetails[*].dimensions.width', byRegex(nonBlank()))
           jsonPath('$.toyDetails[*].dimensions.length', byRegex(nonBlank()))
       }
```

Then the code now thinks that the array should be an isEmpty check:
`assertThatJson(parsedJson).array("['toyDetails']").field("['dimensions']").isEmpty();`

Thoughts?

**Comment by OlgaMaciaszek** [comment]
Hello @mystical44 thanks for this report; could you check if the problem still persists on the latest SCC snapshot?

**Comment by mystical44** [comment]
Hi @OlgaMaciaszek , this still occurs using:

        <dependency>
            <groupId>org.springframework.cloud</groupId>
            <artifactId>spring-cloud-starter-contract-verifier</artifactId>
            <version>2.0.3.BUILD-SNAPSHOT</version>
            <scope>test</scope>
        </dependency>

**Comment by marcingrzejszczak** [comment]
It's the version of the plugin that matters. Can you check `2.1.0.BUILD-SNAPSHOT` and `2.0.3.BUILD-SNAPSHOT` ?

**Comment by mystical44** [comment]
@marcingrzejszczak still no luck using those snapshots.

**Comment by OlgaMaciaszek** [comment]
Ok @mystical44 I'm adding it to our backlog. We will work on it.

**Comment by marcingrzejszczak** [comment]
HI!

I've added such a test case:

```groovy


	@Issue("#797")
	def "should not create an unnecessary empty collection check [#methodBuilderName]"() {
		given:
			Contract contractDsl = Contract.make {
				name("get_new_toy_specs")
				request {
					description("""
            Given: A new toy request is submitted
            When: I receive the response
            Then: I would receive the toy specs
        """)
					method 'GET'
					urlPath ('/toys') {
						queryParameters {
							parameter 'uuid': 'd4d724c4-e36e-4fd2-9baa-af7f5df17399'
						}
					}
				}
				response {
					status 200
					body([
							toyUuid: "d4d724c4-e36e-4fd2-9baa-af7f5df17399",
							toyDescription: [
									name: "Super Whiz Bang Toy",
									stockNum: 1234,
									manufacturer: "Toy Comp",
							],
							toyDetails: [
									[
											inventory: 42,
											description: "Toy of the year!!",
											dimensions: [
													height: 45.8,
													weight: 12.3,
													width: 8.6,
													length: 9.3
											]
									]
							]
					])
					bodyMatchers {
						//toyDescription checks
						jsonPath('$.toyDetails[*].dimensions.weight', byRegex(nonBlank()))
						jsonPath('$.toyDetails[*].dimensions.width', byRegex(nonBlank()))
						jsonPath('$.toyDetails[*].dimensions.length', byRegex(nonBlank()))
					}
					headers {
						contentType(applicationJson())
					}
				}
			}
			MethodBodyBuilder builder = methodBuilder(contractDsl)
			BlockBuilder blockBuilder = new BlockBuilder(" ")
		when:
			builder.appendTo(blockBuilder)
		then:
			String test = blockBuilder.toString()
			SyntaxChecker.tryToCompileWithoutCompileStatic(methodBuilderName, test)
			!test.contains('''assertThatJson(parsedJson).array("['toyDetails']").field("['dimensions']").isEmpty()''')
		and:
			stubMappingIsValidWireMockStub(contractDsl)
		where:
			methodBuilderName                                    | methodBuilder
			"MockMvcSpockMethodBuilder"                          | { Contract dsl -> new MockMvcSpockMethodRequestProcessingBodyBuilder(dsl, properties) }
			"MockMvcJUnitMethodBuilder"                          | { Contract dsl -> new MockMvcJUnitMethodBodyBuilder(dsl, properties) }
			"JaxRsClientSpockMethodRequestProcessingBodyBuilder" | { Contract dsl -> new JaxRsClientSpockMethodRequestProcessingBodyBuilder(dsl, properties) }
			"JaxRsClientJUnitMethodBodyBuilder"                  | { Contract dsl -> new JaxRsClientJUnitMethodBodyBuilder(dsl, properties) }
	}
```

And it's passing for `2.0.3.BUILD-SNAPSHOT`. I'm more than certain that you're using an invalid version of the plugin (not the verifier dependency, but the plugin itself).

**Comment by nicolaimainiero** [comment]
@marcingrzejszczak Looks like you missed the `jsonPath('$.toyDetails[*].dimensions.height', byRegex(nonBlank()))` matcher in your test case. Could you verifiy if still works?

**Comment by marcingrzejszczak** [comment]
Thanks @nicolaimainiero for pointing it out. I haven't noticed it before. Now it's fixed!

**Comment by jasonfbenton** [comment]
We are still seeing this issue in 2.1.0.RELEASE. Here's an example that recreates the issue.

```
package contracts

import org.springframework.boot.autoconfigure.couchbase.CouchbaseProperties
import org.springframework.cloud.contract.spec.Contract

Contract.make {
            description('''
Given: A GET /items request is submitted
When: I receive the response
Then: I would see a list of all items
''')
            request {
                method 'GET'
                url '/items'
                headers {
                    header('Content-Type', 'application/json;charset=UTF-8')
                }
            }
            response {
                status 200
                body(
                        _metadata: $(
                                totalCount: $(anyNumber())
                        ),

                        content: [[
                                          itemId       : '001',
                                          name       : 'foo',
                                          description: 'This is foo'
                                  ], [
                                          itemId       : '002',
                                          name       : 'bar',
                                          description: 'This is bar'
                                  ]]

                )
                bodyMatchers {
                    jsonPath('$.content[0].itemId', byRegex('001'))
                    jsonPath('$.content[0].name', byRegex('foo'))
                    jsonPath('$.content[0].description', byRegex('This is foo'))
                    jsonPath('$.content[1].itemId', byRegex('002'))
                    jsonPath('$.content[1].name', byRegex('bar'))
                    jsonPath('$.content[1].description', byRegex('This is bar'))
                }
                headers {
                    header('Content-Type': 'application/json;charset=UTF-8')
                }
            }
}
```

The test generated from this contract will include the following assertion:
`assertThatJson(parsedJson).array("['content']").isEmpty();`

**Comment by marcingrzejszczak** [comment]
Can you check it against 2.1.1 snapshots?

**Comment by jasonfbenton** [comment]
Sorry for the delayed response. It appears to be fixed in the 2.1.1 snapshot!
